### PR TITLE
[KYUUBI #703] support hudi-0.9.0

### PR DESCRIPTION
**Pull request by simon824**
### _Why are the changes needed?_
Hudi has add Spark Sql Support in version 0.9.0.
This PR add Hudi-0.9.0 dependency and Unit Testing  #703 

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/test/scala/org/apache/kyuubi/operation/BasicHudiJDBCTests.scala
Comment (yanghua) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/operation/BasicHudiJDBCTests.scala:24]: `HudiBasicJDBCTests` sounds better?

File: kyuubi-server/pom.xml
Comment (yanghua) [inline kyuubi-server/pom.xml:138]: IMO, we do not need to follow `${iceberg.name}`, it can be written here directly.

File: pom.xml
Comment (yanghua) [inline pom.xml:1071]: ditto

**Review by yanghua — COMMENTED** [review]
@simon824 Thanks for your contribution, have a quick look, and left some comments.

File: kyuubi-common/src/test/scala/org/apache/kyuubi/operation/BasicHudiJDBCTests.scala
Comment (cfmcgrady) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/operation/BasicHudiJDBCTests.scala:103]:
```suggestion
  }
}
```

**Comment by yaooqinn** [comment]
Hi @simon824, can you take a look at the test failure?

**Comment by yaooqinn** [comment]
```
at org.apache.spark.sql.hudi.command.CreateHoodieTableCommand.createHiveDataSourceTable(CreateHoodieTableCommand.scala:214)
	at org.apache.spark.sql.hudi.command.CreateHoodieTableCommand.createTableInCatalog(CreateHoodieTableCommand.scala:177)
	at org.apache.spark.sql.hudi.command.CreateHoodieTableCommand.run(CreateHoodieTableCommand.scala:73)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
	at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:228)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3687)
```
cc @pan3793, the data lakes are bound to hive metastore sometimes, are we good to use `in-memory` for them all?

**Comment by pan3793** [comment]
> ```
> at org.apache.spark.sql.hudi.command.CreateHoodieTableCommand.createHiveDataSourceTable(CreateHoodieTableCommand.scala:214)
> 	at org.apache.spark.sql.hudi.command.CreateHoodieTableCommand.createTableInCatalog(CreateHoodieTableCommand.scala:177)
> 	at org.apache.spark.sql.hudi.command.CreateHoodieTableCommand.run(CreateHoodieTableCommand.scala:73)
> 	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
> 	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
> 	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
> 	at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:228)
> 	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3687)
> ```
> 
> cc @pan3793, the data lakes are bound to hive metastore sometimes, are we good to use `in-memory` for them all?

Make sense.

**Comment by yaooqinn** [comment]
```
*** RUN ABORTED ***
  java.lang.NoSuchMethodError: org.apache.hadoop.yarn.server.utils.BuilderUtils.newApplicationResourceUsageReport(IILorg/apache/hadoop/yarn/api/records/Resource;Lorg/apache/hadoop/yarn/api/records/Resource;Lorg/apache/hadoop/yarn/api/records/Resource;Ljava/util/Map;Ljava/util/Map;)Lorg/apache/hadoop/yarn/api/records/ApplicationResourceUsageReport;
  at org.apache.hadoop.yarn.server.resourcemanager.RMServerUtils.<clinit>(RMServerUtils.java:493)
  at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.serviceInit(ResourceManager.java:277)
  at org.apache.hadoop.service.AbstractService.init(AbstractService.java:164)
  at org.apache.hadoop.yarn.server.MiniYARNCluster.initResourceManager(MiniYARNCluster.java:348)
  at org.apache.hadoop.yarn.server.MiniYARNCluster.access$200(MiniYARNCluster.java:128)
  at org.apache.hadoop.yarn.server.MiniYARNCluster$ResourceManagerWrapper.serviceInit(MiniYARNCluster.java:497)
  at org.apache.hadoop.service.AbstractService.init(AbstractService.java:164)
  at org.apache.hadoop.service.CompositeService.serviceInit(CompositeService.java:108)
  at org.apache.hadoop.yarn.server.MiniYARNCluster.serviceInit(MiniYARNCluster.java:316)
  at org.apache.hadoop.service.AbstractService.init(AbstractService.java:164)
  ...
```

YARN dep conflicts?

File: pom.xml
Comment (pan3793) [inline pom.xml:1072]: IMO, this package is broken, the `hudi-spark3_2.12` depends on `hudi-spark-common_2.11`, which will cause the Scala library conflict, you can check at
https://mvnrepository.com/artifact/org.apache.hudi/hudi-spark3_2.12/0.9.0

Comment (pan3793) [reply pom.xml:1072]: cc @yanghua

**Comment by pan3793** [comment]
Have a look at `hudi-spark-bundle_2.12`, I'm curious why hudi only relocate part of deps but left others(e.g. curator, prometheus, hbase, rocksdb) with the original package name in the fat jar? It leads to a very high probability of class conflict.
https://github.com/apache/hudi/blob/master/packaging/hudi-spark-bundle/pom.xml

**Comment by pan3793** [comment]
It say Hudi 0.9.0 doesn't support Spark 3.1 https://github.com/apache/hudi/issues/3554

**Comment by simon824** [comment]
> It say Hudi 0.9.0 doesn't support Spark 3.1 [apache/hudi#3554](https://github.com/apache/hudi/issues/3554)

support Spark-2.4.3+ & Spark 3.x, https://hudi.apache.org/docs/quick-start-guide/
Maybe still have some bugs in some aspects.

**Comment by yaooqinn** [comment]
Do we need to test spark 3.0 only for now?

**Comment by simon824** [comment]
> Do we need to test spark 3.0 only for now?

yes

File: kyuubi-server/pom.xml
Comment (yaooqinn) [inline kyuubi-server/pom.xml:149]: the test failure is related to this dep? can we exclude it?

Comment (simon824) [reply kyuubi-server/pom.xml:149]: Hudi needs to depend on hbase.
You mean to exclude it in kyuubi-server module or the whole project?

Comment (yaooqinn) [reply kyuubi-server/pom.xml:149]: the whole project

Comment (ulysses-you) [reply kyuubi-server/pom.xml:149]: > Hudi needs to depend on hbase.

just curious, why hudi depend on hbase ?

Comment (pan3793) [reply kyuubi-server/pom.xml:149]: For index

Comment (ulysses-you) [reply kyuubi-server/pom.xml:149]: ah thanks, it's a good news ..

**Comment by pan3793** [comment]
Congratulation! The CI passed. Thanks to @simon824 for good work. But I need some time to have a deep look at the transitive deps pulled by Hudi before merging it.

File: .github/workflows/master.yml
Comment (yaooqinn) [inline .github/workflows/master.yml:93]: is this already in master？ we can restore this change

Comment (pan3793) [reply .github/workflows/master.yml:93]: I think the `git merge` will auto-resolve this change

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, @simon824, this change LGTM, leave it to @pan3793 for the last comfirmation

**Comment by simon824** [comment]
> thanks, @simon824, this change LGTM, leave it to @pan3793 for the last comfirmation

**Comment by simon824** [comment]
OK, thanks a lot for your review.

**Comment by pan3793** [comment]
Hudi use `HFileBootstrapIndex` by default, I think it's not required for our ut, can we overwrite it and test if it can work without hbase deps?

https://github.com/apache/hudi/blob/release-0.9.0/hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableConfig.java#L144-L147

**Comment by pan3793** [comment]
I still see `hudi-spark_2.12` pulls `jackson-module-scala_2.11` into classpath

**Comment by pan3793** [comment]
Stand on the Spark side, HUDI is a data source implementation, or just call it a plugin. 

Thus, for the transitive dependencies of HUDI, like Hadoop, Curator, Parquet, Avro, ORC, Hive, we must **RESPECT** the Spark bundled versions, as Kyuubi support multi Spark versions, we need to separate those deps version by different Spark versions using maven profiles. e.g. use parquet 1.10.0 for Spark 3.0/3.1, but use parquet 1.12.0 for Spark 3.2(coming soon).

Besides on Spark, we should also respect the Kyuubi deps version, assume you are using IDEA, please open the `External Libraries`, there are lots of jars which has multi versions that exist in the classpath, which means a high risk of class conflicts.

**Comment by yaooqinn** [comment]
it worths another PR to solve, this PR has met its original goal. thanks, merged to master.

**Comment by simon824** [comment]
> Hudi uses `HFileBootstrapIndex` by default, I don't think it's required for our ut, can we overwrite it and test if it can work without hbase deps?
> 
> https://github.com/apache/hudi/blob/release-0.9.0/hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableConfig.java#L144-L147

ok，i will test and fix it.

**Pull request closed without merging**
